web-api/src/personalization/stateless.rs

```diff
@@ -95,26 +97,38 @@ pub(super) async fn load_history(
     storage: &impl storage::Document,
     history: Vec<HistoryEntry>,
 ) -> Result<Vec<LoadedHistoryEntry>, Error> {
-    let mut loaded =
-        storage::Document::get_interacted(storage, history.iter().map(|entry| &entry.id))
-            .await?
-            .into_iter()
-            .map(|document| (document.id, (document.embedding, document.tags)))
-            .collect::<HashMap<_, _>>();
+    let history = history
+        .into_iter()
+        .map(|entry| {
+            // TODO[pmk/soon] properly support history of documents with multiple snippets
+            let id = match entry.id {
+                SnippetOrDocumentId::SnippetId(id) => id,
+                SnippetOrDocumentId::DocumentId(id) => SnippetId::new(id, 0),
+            };
+            (id, entry.timestamp)
+        })
+        .collect::<HashMap<_, _>>();
 
-    Ok(history
+    let loaded = storage::Document::get_snippets_for_interaction(storage, history.keys())
+        .await?
+        .into_iter();
+
+    Ok(loaded
         .into_iter()
-        // filter ignores documents which don't exist in our database (i.e. have
-        // been deleted)
-        .filter_map(|HistoryEntry { id, timestamp }| {
-            loaded
-                .remove(&id)
-                .map(|(embedding, tags)| LoadedHistoryEntry {
+        .map(
+            |SnippetForInteraction {
+                 id,
+                 embedding,
+                 tags,
+             }| {
+                let timestamp = history[&id];
```

Comment:
As `loaded ⊆ history`, this shouldn't be able to panic. Which is why I iterate over loaded and get from history instead of iterating over history and getting from loaded.